engine/kv_engine_cleaner.hpp

```diff
@@ -80,48 +80,114 @@ class Cleaner {
   Cleaner(KVEngine* kv_engine, int64_t max_cleaner_threads)
       : kv_engine_(kv_engine),
         max_thread_num_(max_cleaner_threads),
+        min_thread_num_(1),
         close_(false),
         start_slot_(0),
-        live_thread_num_(0),
-        workers_(max_cleaner_threads) {
-    for (size_t thread_id = 0; thread_id < max_thread_num_; ++thread_id) {
-      idled_workers_.push_back(thread_id);
+        active_clean_workers_(0),
+        main_worker_([this]() { this->mainWork(); }),
+        clean_workers_(max_thread_num_ - 1 /*1 for main worker*/) {
+    for (auto& w : clean_workers_) {
+      w.Init([this]() { this->cleanWork(); });
     }
   }
 
-  ~Cleaner() { CloseAllWorkers(); }
+  ~Cleaner() { Close(); }
 
-  void StartClean();
-  void CloseAllWorkers() {
+  void Start() {
+    if (!close_ && min_thread_num_ > 0) {
+      main_worker_.Run();
+    }
+  }
+
+  void Close() {
     close_ = true;
-    for (size_t i = 0; i < workers_.size(); ++i) {
-      if (workers_[i].worker.joinable()) {
-        workers_[i].finish = true;
-        workers_[i].worker.join();
-      }
+    main_worker_.Join();
+    for (auto& w : clean_workers_) {
+      w.Join();
     }
   }
-  void AdjustThread(size_t advice_thread_num);
-  size_t ActiveThreadNum() { return live_thread_num_.load(); }
+
+  void AdjustCleanWorkers(size_t advice_workers_num);
+
+  size_t ActiveThreadNum() { return active_clean_workers_.load() + 1; }
 
   double SearchOutdatedCollections();
   void FetchOutdatedCollections(PendingCleanRecords& pending_clean_records);
 
  private:
-  struct ThreadWorker {
-    std::atomic_bool finish{true};
-    std::thread worker;
+  class Worker {
+   public:
+    ~Worker() { Join(); }
+
+    Worker(std::function<void()> f) { Init(f); }
+
+    Worker() = default;
+
+    void Init(std::function<void()> f) {
+      Join();
+      func = f;
+      std::unique_lock<SpinMutex> ul(spin);
+      join_ = false;
+      keep_work_ = false;
+
+      auto work = [&]() {
+        while (true) {
+          {
+            std::unique_lock<SpinMutex> ul(spin);
```

Comment:
Dead lock?

Reply:
no, this is thread's work, this function Init will return

Reply:
This will be unlocked when it comes to `cv.wait(ul)`, or when it is out of scope. So it won't lock here for a long time.